lib/ansible/modules/source_control/gitlab_deploy_key.py

```diff
@@ -94,140 +107,216 @@
+  type: dict
 '''
 
+import os
+import re
 
-import json
+try:
+    import gitlab
+    HAS_GITLAB_PACKAGE = True
+except Exception:
+    HAS_GITLAB_PACKAGE = False
 
 from ansible.module_utils.basic import AnsibleModule
-from copy import deepcopy
-from ansible.module_utils.gitlab import request
-
-
-def _list(module, api_url, project, access_token, private_token):
-    path = "/deploy_keys"
-    return request(module, api_url, project, path, access_token, private_token)
-
-
-def _find(module, api_url, project, key, access_token, private_token):
-    success, data = _list(module, api_url, project, access_token, private_token)
-    if success:
-        for i in data:
-            if i["key"] == key:
-                return success, i
-        return success, None
-    return success, data
-
-
-def _publish(module, api_url, project, data, access_token, private_token):
-    path = "/deploy_keys"
-    method = "POST"
-    if 'id' in data:
-        path += "/%s" % str(data["id"])
-        method = "PUT"
-    data = deepcopy(data)
-    data.pop('id', None)
-    return request(module, api_url, project, path, access_token, private_token, json.dumps(data, sort_keys=True), method)
-
-
-def _delete(module, api_url, project, key_id, access_token, private_token):
-    path = "/deploy_keys/%s" % str(key_id)
-    return request(module, api_url, project, path, access_token, private_token, method='DELETE')
-
-
-def _are_equivalent(input, existing):
-    for key in ['title', 'key', 'can_push']:
-        if key in input and key not in existing:
-            return False
-        if key not in input and key in existing:
-            return False
-        if not input[key] == existing[key]:
+from ansible.module_utils._text import to_native
+
+from ansible.module_utils.gitlab import findProject
+
+
+class GitLabDeployKey(object):
+    def __init__(self, module, gitlab_instance):
+        self._module = module
+        self._gitlab = gitlab_instance
+        self.deployKeyObject = None
+
+    '''
+    @param project Project object
+    @param key_title Title of the key
+    @param key_key String of the key
+    @param key_can_push Option of the deployKey
+    @param options Deploy key options
+    '''
+    def createOrUpdateDeployKey(self, project, key_title, key_key, options):
+        changed = False
+
+        # Because we have already call existsDeployKey in main()
+        if self.deployKeyObject is None:
+            deployKey = self.createDeployKey(project, {
+                'title': key_title,
+                'key': key_key,
+                'can_push': options['can_push']})
+            changed = True
+        else:
+            changed, deployKey = self.updateDeployKey(self.deployKeyObject, {
+                'can_push': options['can_push']})
+
+        self.deployKeyObject = deployKey
+        if changed:
+            if self._module.check_mode:
+                self._module.exit_json(changed=True, result="DeployKey should have been updated.")
```

Comment:
Please use 'msg' for returned messages, this is a standardized return value.

Reply:
Will be fix in the next commit